util/fp_finder.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2025 OWASP Core Rule Set Project
+// SPDX-License-Identifier: Apache-2.0
+
+package util
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"regexp"
+	"slices"
+	"strings"
+
+	"github.com/coreruleset/crs-toolchain/v2/utils"
+)
+
+type FpFinderError struct{}
+
+const dictionaryURLFormat = "https://raw.githubusercontent.com/dwyl/english-words/%s/%s"
+const dictionaryBaseFileName = "words_alpha.txt"
+const minSize = 3
+
+func (t *FpFinderError) Error() string {
+	return "FpFinder error"
+}
+
+type FpFinder struct{}
+
+func NewFpFinder() *FpFinder {
+	return &FpFinder{}
+}
+
+func (t *FpFinder) FpFinder(inputFilePath string, extendedDictionaryFilePath string, englishDictionaryCommitHash string) error {
+	// Get the dictionary path in ~/.crs-toolchain
+	dictionaryFileName := fmt.Sprintf("%s-%s", englishDictionaryCommitHash, dictionaryBaseFileName)
+	dictionaryPath, err := utils.GetCacheFilePath(dictionaryFileName)
+	if err != nil {
+		logger.Fatal().Err(err).Msgf("Error getting dictionary path: %v", err)
+	}
+
+	// Check if the dictionary exists, if not, download it
+	if _, err := os.Stat(dictionaryPath); os.IsNotExist(err) {
+		logger.Debug().Msg("Dictionary file not found. Downloading...")
+		dictionaryURL := fmt.Sprintf(dictionaryURLFormat, englishDictionaryCommitHash, dictionaryBaseFileName)
+		if err := utils.DownloadFile(dictionaryPath, dictionaryURL); err != nil {
+			logger.Fatal().Err(err).Msg("Failed to download dictionary")
+		}
+		logger.Debug().Msg("Download complete.")
+	} else {
+		logger.Debug().Msg("Dictionary file found, skipping download.")
+	}
+
+	// Load dictionary into memory
+	englishDict, err := t.loadDictionary(dictionaryPath, minSize)
+	if err != nil {
+		logger.Fatal().Err(err).Msg("Failed to load english dictionary")
+	}
+
+	var dict map[string]struct{}
+	if extendedDictionaryFilePath != "" {
+		extendedDict, err := t.loadDictionary(extendedDictionaryFilePath, 0)
+		if err != nil {
+			logger.Fatal().Err(err).Msg("Failed to load extended dictionary")
+		}
+
+		// Add words from extendedDictionary
+		dict = t.mergeDictionaries(englishDict, extendedDict)
+	} else {
+		dict = englishDict
+	}
+
+	// Load input file into memory
+	inputFile, err := t.loadFileContent(inputFilePath)
+	if err != nil {
+		logger.Fatal().Err(err).Msg("Failed to load input file")
+	}
+
+	// Filter words not in dictionary, remove duplicates, and sort alphabetically
+	filteredWords := t.filterContent(inputFile, dict, minSize)
+
+	// Remove adjacent duplicate words from the sorted list
+	filteredWords = slices.Compact(filteredWords)
+
+	for _, str := range filteredWords {
+		fmt.Println(str)
+	}
+
+	return nil
+}
+
+func (t *FpFinder) loadDictionary(path string, minSize int) (map[string]struct{}, error) {
```

Comment:
This method mostly duplicates `loadFileContent()`. Consider calling `loadFileContent()` and building the dictionary from the returned data.